### PR TITLE
Add a WebJar Locator extension that duplicates webjar locator behavior from Spring Boot

### PR DESCRIPTION
**Pull request by Christopher-Chianelli**
Currently, if you use webjars in your application, you need to specify the version in the URL. Spring-Boot allows you to not specify the webjar version in the URL by adding the `webjar-locator` dependency. I propose we replicate this functionality with a Quarkus extension.

 Side note: is it WebJar Locator or Webjars Locator? The artifactId and name from Maven Central is inconsistent (see https://mvnrepository.com/artifact/org.webjars/webjars-locator/0.39 and https://github.com/webjars/webjars-locator-core/blob/master/pom.xml).

To add: tests (which I guess would be testing the handler for given root paths and webjar name to version map)

### CODE REVIEW

File: extensions/webjars-locator/deployment/pom.xml
Comment (stuartwdouglas) [inline extensions/webjars-locator/deployment/pom.xml]: This should be managed from the runtime bom (and be project.version).

File: extensions/webjars-locator/deployment/pom.xml
Comment (stuartwdouglas) [inline extensions/webjars-locator/deployment/pom.xml]: This should be managed from the deployment bom.

File: extensions/webjars-locator/runtime/src/main/resources/META-INF/quarkus-extension.yaml
Comment (stuartwdouglas) [inline extensions/webjars-locator/runtime/src/main/resources/META-INF/quarkus-extension.yaml]: This guide should be empty if there is no guide for this, although it would be great if you could add a note about this to the HTTP reference instead.

File: extensions/webjars-locator/deployment/src/main/java/io/quarkus/webjar/locator/deployment/WebJarLocatorStandaloneBuildStep.java
Comment (stuartwdouglas) [inline extensions/webjars-locator/deployment/src/main/java/io/quarkus/webjar/locator/deployment/WebJarLocatorStandaloneBuildStep.java]: Why is this not needed for Servlet?

File: extensions/webjars-locator/deployment/src/main/java/io/quarkus/webjar/locator/deployment/WebJarLocatorStandaloneBuildStep.java
Comment (stuartwdouglas) [inline extensions/webjars-locator/deployment/src/main/java/io/quarkus/webjar/locator/deployment/WebJarLocatorStandaloneBuildStep.java]: I don't think you actually need WebJarLocatorConfigBuildItem, it looks like this could just be a single build step?

**Review by stuartwdouglas — CHANGES_REQUESTED** [review]

File: extensions/webjars-locator/deployment/src/main/java/io/quarkus/webjar/locator/deployment/WebJarLocatorStandaloneBuildStep.java
Comment (geoand) [reply extensions/webjars-locator/deployment/src/main/java/io/quarkus/webjar/locator/deployment/WebJarLocatorStandaloneBuildStep.java]: Can you please elaborate on this? I am sure you have some good technical reason in mind, but I am also sure that Quarkus has a lot up it's sleeve to lift any such restrictions

File: extensions/webjars-locator/deployment/src/main/java/io/quarkus/webjar/locator/deployment/WebJarLocatorStandaloneBuildStep.java
Comment (geoand) [inline extensions/webjars-locator/deployment/src/main/java/io/quarkus/webjar/locator/deployment/WebJarLocatorStandaloneBuildStep.java]: The name doesn't seem really fitting :)

**Comment by geoand** [comment]
@Christopher-Chianelli Thanks for the contribution!

When adding tests for this, it is important to keep in mind that we will need tests for native mode and also for dev mode to ensure that all our bases are covered.

**Comment by geoand** [comment]
It seems that webjars uses https://github.com/classgraph/classgraph, so I am thinking that we definitely need to see how it's usage would affect the time of dev-mode restart.

File: extensions/webjars-locator/deployment/pom.xml
Comment (ge0ffrey) [reply extensions/webjars-locator/deployment/pom.xml]: see https://github.com/quarkusio/quarkus/blob/master/bom/runtime/pom.xml

File: extensions/webjars-locator/deployment/src/main/java/io/quarkus/webjar/locator/deployment/WebJarLocatorStandaloneBuildStep.java
Comment (ge0ffrey) [reply extensions/webjars-locator/deployment/src/main/java/io/quarkus/webjar/locator/deployment/WebJarLocatorStandaloneBuildStep.java]: Isn't the reverse needed? if servlet compability is _not_ present, return?

File: extensions/webjars-locator/runtime/src/main/resources/META-INF/quarkus-extension.yaml
Comment (ge0ffrey) [reply extensions/webjars-locator/runtime/src/main/resources/META-INF/quarkus-extension.yaml]: @Christopher-Chianelli How about adding a really short guide for webjars here: https://github.com/quarkusio/quarkus/tree/master/docs/src/main/asciidoc ?

Also, an integration test here to prove it works natively would be great:
https://github.com/quarkusio/quarkus/tree/master/integration-tests

File: extensions/webjars-locator/deployment/src/main/java/io/quarkus/webjar/locator/deployment/WebJarLocatorStandaloneBuildStep.java
Comment (Christopher-Chianelli) [reply extensions/webjars-locator/deployment/src/main/java/io/quarkus/webjar/locator/deployment/WebJarLocatorStandaloneBuildStep.java]: I based this extension on the resteasy-extension: https://github.com/quarkusio/quarkus/blob/0d29a66fe75b43dc9914c738e4ceb876272d0d74/extensions/resteasy/deployment/src/main/java/io/quarkus/resteasy/deployment/ResteasyStandaloneBuildStep.java#L62-L103. I figured if `RESTeasy` has it for a reason, this extension will probably have it for the same reason. It may be entirely unneeded.

File: extensions/webjars-locator/deployment/src/main/java/io/quarkus/webjar/locator/deployment/WebJarLocatorStandaloneBuildStep.java
Comment (Christopher-Chianelli) [reply extensions/webjars-locator/deployment/src/main/java/io/quarkus/webjar/locator/deployment/WebJarLocatorStandaloneBuildStep.java]: Will there be any performance downside though? Or are all BuildStep (regardless if recorded or not) rerun at runtime? (the idea behind having it in two separate methods is to put the possibly computational expensive/classpath scanning at buildtime, and use the saved Map when we start the application)

File: extensions/webjars-locator/deployment/src/main/java/io/quarkus/webjar/locator/deployment/WebJarLocatorStandaloneBuildStep.java
Comment (Christopher-Chianelli) [reply extensions/webjars-locator/deployment/src/main/java/io/quarkus/webjar/locator/deployment/WebJarLocatorStandaloneBuildStep.java]: The name was copied from resteasy extension. How about `findWebjarsAndRootPath`?

File: extensions/webjars-locator/deployment/src/main/java/io/quarkus/webjar/locator/deployment/WebJarLocatorStandaloneBuildStep.java
Comment (geoand) [reply extensions/webjars-locator/deployment/src/main/java/io/quarkus/webjar/locator/deployment/WebJarLocatorStandaloneBuildStep.java]: Build Steps are build-time only. What gets run at runtime are the recorders (that are associated with a build step).
So there is no reason to have different build steps for what you are doing.

File: extensions/webjars-locator/deployment/src/main/java/io/quarkus/webjar/locator/deployment/WebJarLocatorStandaloneBuildStep.java
Comment (geoand) [reply extensions/webjars-locator/deployment/src/main/java/io/quarkus/webjar/locator/deployment/WebJarLocatorStandaloneBuildStep.java]: Sure :)

**Comment by Christopher-Chianelli** [comment]
Note: Test fails with
```
Caused by: java.lang.Error: Unresolved compilation problem: 
	The type sun.misc.Unsafe cannot be resolved. It is indirectly referenced from required .class files

	at io.quarkus.runtime.ExecutorRecorder.<init>(ExecutorRecorder.java:1)
	at io.quarkus.deployment.steps.ThreadPoolSetup$createExecutor7.deploy_0(ThreadPoolSetup$createExecutor7.zig:59)
	at io.quarkus.deployment.steps.ThreadPoolSetup$createExecutor7.deploy(ThreadPoolSetup$createExecutor7.zig:36)
	at io.quarkus.runner.ApplicationImpl.doStart(ApplicationImpl.zig:106)
	... 44 more
```
For me. Maybe `webjar-locator` uses it somewhere (meaning we'll need to implement it main functionality without it)

**Comment by geoand** [comment]
> Note: Test fails with
> 
> ```
> Caused by: java.lang.Error: Unresolved compilation problem: 
> 	The type sun.misc.Unsafe cannot be resolved. It is indirectly referenced from required .class files
> 
> 	at io.quarkus.runtime.ExecutorRecorder.<init>(ExecutorRecorder.java:1)
> 	at io.quarkus.deployment.steps.ThreadPoolSetup$createExecutor7.deploy_0(ThreadPoolSetup$createExecutor7.zig:59)
> 	at io.quarkus.deployment.steps.ThreadPoolSetup$createExecutor7.deploy(ThreadPoolSetup$createExecutor7.zig:36)
> 	at io.quarkus.runner.ApplicationImpl.doStart(ApplicationImpl.zig:106)
> 	... 44 more
> ```
> 
> For me. Maybe `webjar-locator` uses it somewhere (meaning we'll need to implement it main functionality without it)

Did it fail with in a JVM test or native test?

**Comment by Christopher-Chianelli** [comment]
@geoand I think it failed with a JVM test (ran from maven, fails on both Java 11 and Java 8); the test use `QuarkusUnitTest`.

**Comment by geoand** [comment]
Thanks for the information @Christopher-Chianelli. Yes that would be a JVM mode test then that failed.

**Comment by Christopher-Chianelli** [comment]
Hmm... I think my IDE (Flatpak Eclipse) might be doing weird magic behind the scenes. If I build Quarkus from scratch using maven, I get an expectation error (which means it was able to start Quarkus). If I open my IDE (without doing any changes), I get `java.lang.RuntimeException: java.lang.RuntimeException: java.lang.RuntimeException: Failed to load steps from class io.quarkus.vertx.http.deployment.HttpSecurityProcessor`. So look like Eclipse's build seems break Quarkus and corrupts my .m2?

**Comment by Christopher-Chianelli** [comment]
Update: `quarkus-webjar-locator` unit test now actually works. Turns out you need to add forced dependencies into `pom.xml` test's scope for webjars. Also Eclipse automatic build can be troublesome (as in, breaks Quarkus). Working on native test now.

**Comment by Christopher-Chianelli** [comment]
Update: Add integration and native test for `quarkus-webjar-locator`.

**Comment by Christopher-Chianelli** [comment]
Update: Add `dev-mode` test for the `webjars-locator` extension.
Notes:

- The extension does not provide web server capabilities, and is reliant on there being
an extension that add web server capabilities. I use resteasy in the tests, but it *should*
work with any extension that adds webserver capabilities.

**Comment by Christopher-Chianelli** [comment]
Java 8 build failed at `quarkus-smallrye-reactive-messaging-amqp-deployment`:
`AmqpDevModeTest.testCodeUpdate » Composite 2 exceptions occurred.` 

This appears to be the root cause:
```
2020-04-14 19:05:22,446 INFO  [io.sma.rea.mes.ext.MediatorManager] (Quarkus Main Thread) Cancel subscriptions
2020-04-14 19:05:22,449 INFO  [io.quarkus] (Quarkus Main Thread) Quarkus stopped in 0.053s
java.lang.IllegalStateException: Arc context propagation was attempted but the container is not running.
	at io.quarkus.arc.runtime.context.ArcContextProvider.lambda$currentContext$2(ArcContextProvider.java:35)
	at io.smallrye.context.impl.ActiveContextState.<init>(ActiveContextState.java:18)
	at io.smallrye.context.impl.CapturedContextState.begin(CapturedContextState.java:35)
	at io.smallrye.context.SmallRyeThreadContext.lambda$withContext$0(SmallRyeThreadContext.java:213)
	at io.smallrye.context.propagators.rxjava2.ContextPropagatorOnFlowableCreateAction$ContextCapturerFlowable.onNext(ContextPropagatorOnFlowableCreateAction.java:50)
	at io.reactivex.internal.operators.flowable.FlowableFlatMap$MergeSubscriber.drainLoop(FlowableFlatMap.java:506)
```

Doesn't appear related to the PR.

**Comment by geoand** [comment]
Can you do me a favor and rebase onto master?

**Comment by Christopher-Chianelli** [comment]
Sure, (I prefer merging master back into the branch since rebases need to be forced-push, cause comments to go missing, amd mess up history). One sec

**Comment by Christopher-Chianelli** [comment]
@geoand Done

**Comment by geoand** [comment]
@ia3andy would you like to review this as well?

File: extensions/webjars-locator/deployment/src/test/java/io/quarkus/webjar/locator/test/WebJarLocatorDevModeTest.java
Comment (geoand) [inline extensions/webjars-locator/deployment/src/test/java/io/quarkus/webjar/locator/test/WebJarLocatorDevModeTest.java:14]: Actually I think this test doesn't actually test dev mode changes (unless I am missing something). The point of the `QuarkusDevModeTest` is to make it easy to change application code and/or configuration and make sure the application works properly with the new changes.

Does that make sense?

File: docs/src/main/asciidoc/http-reference.adoc
Comment (ge0ffrey) [inline docs/src/main/asciidoc/http-reference.adoc:43]: I'd also add the jquery webjar itself here in the pom example too, for illustration:

```
    <dependency>
      <groupId>org.webjars</groupId>
      <artifactId>jquery</artifactId>
    </dependency>
```

**Review by ge0ffrey — APPROVED** [review]

File: extensions/webjars-locator/deployment/src/test/java/io/quarkus/webjar/locator/test/WebJarLocatorDevModeTest.java
Comment (Christopher-Chianelli) [reply extensions/webjars-locator/deployment/src/test/java/io/quarkus/webjar/locator/test/WebJarLocatorDevModeTest.java:14]: True, issue is you cannot change dependencies for `QuarkusDevModeTest` (or if there is I couldn't find it), which is basically the only way of modifying the behaviour of this extension (ex: adding the jquery webjar dependency, removing the moment webjar dependency, etc. would be the ideal dev mode test).

Comment (Christopher-Chianelli) [reply extensions/webjars-locator/deployment/src/test/java/io/quarkus/webjar/locator/test/WebJarLocatorDevModeTest.java:14]: This test as it currently is basically make sure nothing odd happens when you run in dev mode.

Comment (geoand) [reply extensions/webjars-locator/deployment/src/test/java/io/quarkus/webjar/locator/test/WebJarLocatorDevModeTest.java:14]: I understand that, but we need to make sure that in an application that contains the extension, restart works properly.

So I propose you do the following: add a dummy JAX-RS resource (like other dev mode tests do), and change a string value in it.
Then hit the endpoints again to ensure that everything still works after the restart.

**Comment by geoand** [comment]
Thanks a lot for this work @Christopher-Chianelli!

Mind squashing the commits into a single one?

**Comment by Christopher-Chianelli** [comment]
@geoand Done!

**Comment by geoand** [comment]
@stuartwdouglas your concerns seem to be addressed, right?

File: extensions/webjars-locator/runtime/src/main/java/io/quarkus/webjar/locator/runtime/WebJarLocatorHandler.java
Comment (stuartwdouglas) [inline extensions/webjars-locator/runtime/src/main/java/io/quarkus/webjar/locator/runtime/WebJarLocatorHandler.java:12]: These should be parameters on the getHandler() method, recorders are not really supposed to be stateful.

File: extensions/webjars-locator/runtime/src/main/java/io/quarkus/webjar/locator/runtime/WebJarLocatorHandler.java
Comment (stuartwdouglas) [inline extensions/webjars-locator/runtime/src/main/java/io/quarkus/webjar/locator/runtime/WebJarLocatorHandler.java:10]: This should be WebJarLocatorRecorder to be in keeping with the rest of the code base

**Review by stuartwdouglas — CHANGES_REQUESTED** [review]
There is a little bit of minor stuff that I missed the first time around, otherwise looks good.

**Comment by Christopher-Chianelli** [comment]
@stuartwdouglas Merged your suggested changes.

**Review by stuartwdouglas — APPROVED** [review]

**Comment by geoand** [comment]
Sorry, can you please rebase the Pull Request onto master and fix the conflict? Once that is done we can merge it.

**Review by geoand — APPROVED** [review]

**Comment by gsmet** [comment]
Found a few minor infelicities (mostly formatting issues). I will amend the commit, rebase, fix the conflict and get this in.

Please don't merge :).

**Comment by geoand** [comment]
Seems like it needs a rebase now.

**Comment by Christopher-Chianelli** [comment]
@gsmet are you going to rebase it or should I?

**Comment by emmanuelbernard** [comment]
@Christopher-Chianelli If you don't, it's likely Guillaume will but it would be much appreciated if you do :)

**Comment by Christopher-Chianelli** [comment]
@geoand , @emmanuelbernard rebased

**Comment by geoand** [comment]
@gsmet did you end up making the changes you wanted to make?

**Comment by gsmet** [comment]
I made all the necessary adjustments in a separate commit and also rebased everything.

In passing, I found out that there was no native tests for this feature and adding one make it fail. I get a 404. I tried adding the resources with `quarkus.native.resources.includes=META-INF/resources/webjars/**` in the application.properties but it didn't work (we would need a proper build item for that anyway but that was to test if it was just resources missing).

I don't have the time to debug that right now and see what's necessary so I let someone else take the lead on this.

**Comment by Christopher-Chianelli** [comment]
@gsmet I will investigate

**Comment by gsmet** [comment]
@stuartwdouglas so we have an issue with this classgraph library. Looks like it doesn't play well with our class loader infrastructure. While things work well when in tests, it doesn't work as soon as you build a real runner jar: it doesn't find the webjar.

I tried to push the TCCL to it but didn't work well: at least the behavior was consistent, it doesn't find anything in all cases :).

We can push the resources to Webjars Locator instead of using classgraph if we have it but I don't know if we have an efficient way to find the artifacts in which we have `META-INF/resources/webjars/` and then scan all the resources here without going through the jar entries. We don't have any classes, just plain resources.

Any idea pops up?

Wondering if we could mimic something similar to what is done with `AdditionalApplicationArchiveMarkerBuildItem`? Was also wondering if it could be something useful in general e.g. have the hability to declare a marker and a tag and be able to get all the archives with a given tag easily.

Note that I might miss something obvious and maybe we already have everything in place to do that!

**Comment by Christopher-Chianelli** [comment]
I got it to work with the runner Jar but not the native image using this (and remove the version map):
```
 event.vertx().fileSystem().readDir("META-INF/resources/webjars/" + webjar.toString(), readDir -> {
    // Find version in file list, and use it to resolve path
});
```

When it is run as a native image, the directory is empty. When run as a runner jar, the directory has the version of the webjar.

**Comment by stuartwdouglas** [comment]
I have changed this to just resolve the versions at build time.

**Comment by Christopher-Chianelli** [comment]
@stuartwdouglas your changes LGTM; All tests passed in native mode.

**Comment by gsmet** [comment]
The test failures are related. The deployment tests are failing with 404 with JDK 8. And on Windows too.

**Comment by Christopher-Chianelli** [comment]
I added a commit that use File.separator and the Path API in order to get the Windows directory to see if that changes anything

**Comment by Christopher-Chianelli** [comment]
Found out why Windows was failing: we were using
```
String resolvedPath = webjarsRootUrl + webjar.toString() + "/" +
                            version + "/" + webjar.relativize(rest).toString();`
```
for redirection, but `webjar.relativize(rest).toString()` use `\` instead of `/` on Windows. To fix this, I added this small method to do the `webjar.relativize(rest).toString()` bit:

```
    private static String convertToRelativeUrl(Path root, Path rest) {
        Path relativePath = root.relativize(rest);
        StringBuilder url = new StringBuilder();
        for (Path part : relativePath) {
            url.append("/");
            try {
                url.append(URLEncoder.encode(part.toString(), "UTF-8"));
            } catch (UnsupportedEncodingException e) {
                // Shouldn't happen, "UTF-8 should be available in all JVM's
                throw new IllegalStateException(e);
            }
        }
        return url.toString();
    }
```

**Comment by Christopher-Chianelli** [comment]
@gsmet Summary of failed checks:

- Windows: Fails due to
```
[ERROR] Errors: 
[ERROR]   JarRunnerIT.testThatFastJarFormatWorks:106 � ConditionTimeout Condition with i...
```
WebJar Locator extensions tests passed; WebJar Locator IT was not run due to the above failure.

- Native Tests - Meassaging: Fails due to
```
 [INFO] Running io.quarkus.it.kafka.streams.KafkaStreamsITCase
[ERROR] Tests run: 1, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 0.019 s <<< FAILURE! - in io.quarkus.it.kafka.streams.KafkaStreamsITCase
[ERROR] io.quarkus.it.kafka.streams.KafkaStreamsITCase  Time elapsed: 0.017 s  <<< ERROR!
org.junit.platform.commons.JUnitException: @Inject is not supported in NativeImageTest tests. Offending field is io.quarkus.it.kafka.streams.KafkaStreamsTest.stateListener

[INFO] 
[INFO] Results:
[INFO] 
[ERROR] Errors: 
[ERROR]   KafkaStreamsITCase » JUnit @Inject is not supported in NativeImageTest tests. ...
[INFO] 
[ERROR] Tests run: 1, Failures: 0, Errors: 1, Skipped: 0
```

- Native Tests - Misc 4 (which contains WebJar Locator native tests) was cancelled during the build phase of the executable (it passes on my machine; guess: OOM or something of the like) (the JAR version of the test ran and passed, native untested).

Situlation this leaves us in regards for this PR: non-native tests pass on JDK 8 and higher on Linux and Windows; native tests are untested (pass on my machine).

**Comment by Christopher-Chianelli** [comment]
There might be an issue if JDK 8 can also create native images, as it uses a workaround that does not work with native images, but works with an executable JAR. If native images can only be created on JDK 11 and above, there should be no issues.

**Comment by stuartwdouglas** [comment]
oops, I tried to fix the test failures but pushed the results to my branch out of habit, instead of this PR. I have created a new PR with my changes for testing: https://github.com/quarkusio/quarkus/pull/9392

Note that the windows path stuff should not matter, / is still a valid separator on windows.

**Comment by geoand** [comment]
So I assume we should close this in favor of #9392 right?

**Comment by stuartwdouglas** [comment]
Lets see what CI has to say, I did not notice there was a conflict so it did not run.

**Comment by Christopher-Chianelli** [comment]
It does matter in URL's: what was happening was the url '/webjar/path/to/file' was being converted to '/webjar\path\to\file', which is a problem

**Comment by Christopher-Chianelli** [comment]
If CI passes for #9392, I am in favour of closing this PR for  #9392

**Comment by gsmet** [comment]
Superseded by https://github.com/quarkusio/quarkus/pull/9392 .

**Pull request closed without merging**
